File: HW1/javaApp/src/main/java/io/rienel/domain/Dog.java
package io.rienel.domain;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * todo armotozov
 *
 * @author Artem Motozov
 * @since 2022.09.19
 */
public class Dog extends Animal<Cat>{

    private final List<Cat> catEaten;

    private String ownerName;

    public Dog(final String name,
               final Double weight,
               final Double fullHeight,
               final Double fullLength,
               final Double fullWidth,
               final String ownerName) {
        super(name, weight, AnimalType.DOG, fullHeight, fullLength, fullWidth);
        catEaten = new ArrayList<>();
        this.ownerName = ownerName;
    }

    public List<Cat> getCatEaten() {
        return catEaten;
    }

    public String getOwnerName() {
        return ownerName;
    }

    public void setOwnerName(String ownerName) {
        this.ownerName = ownerName;
    }

    @Override
    public void setName(String name) {
        Objects.requireNonNull(name);
        super.setName(name);
    }

    @Override
    public void setWeight(Double weight) {
        Objects.requireNonNull(weight);
        if (weight < 2.0 || weight > 15.0) {
            throw new IllegalArgumentException("Illegal weight");
        }
        super.setWeight(weight);
    }

    @Override
    public void setFullHeight(Double fullHeight) {
        Objects.requireNonNull(fullHeight);
        if (fullHeight < 2.0 || fullHeight > 10.0) {
            throw new IllegalArgumentException("Illegal fullHeight");
        }
        super.setFullHeight(fullHeight);
    }

    @Override
    public void setFullLength(Double fullLength) {
        Objects.requireNonNull(fullLength);
        if (fullLength < 2.0 || fullLength > 10.0) {
            throw new IllegalArgumentException("Illegal fullLength");
        }
        super.setFullLength(fullLength);
    }

    @Override
    public void setFullWidth(Double fullWidth) {
        Objects.requireNonNull(fullWidth);
        if (fullWidth < 2.0 || fullWidth > 10.0) {
            throw new IllegalArgumentException("Illegal fullWidth");
        }
        super.setFullWidth(fullWidth);
    }

    @Override
    public Boolean eat(Cat food) {
        Objects.requireNonNull(food);

        if (catEaten.size() > 10) {
            System.out.println("DOG: ENOUGH");
            return false;
        }
        catEaten.add(food);
        return true;
    }

    @Override
    public String speak() {
        return "DOG: BARK";
    }
}

File: HW1/kotlinApp/src/main/kotlin/io/rienel/domain/Dog.kt
package io.rienel.domain

import io.rienel.domain.interfaces.Feedable
import io.rienel.domain.interfaces.Speekable

/**
 * todo armotozov
 *
 * @author Artem Motozov
 * @since 2022.09.19
 */
class Dog(
    name: String,
    weight: Double,
    fullHeight: Double,
    fullLength: Double,
    fullWidth: Double,
    val catEaten: MutableList<Cat>,
    var ownerName: String
) : Animal(
    name,
    weight,
    AnimalType.DOG,
    fullHeight,
    fullLength,
    fullWidth
), Feedable<Cat>, Speekable {

    override var name: String
        get() = super.name
        set(value) {
            assert(value.isNotEmpty())
            super.name = value
        }
    override var weight: Double
        get() = super.weight
        set(value) {
            assert(value in 3.0 .. 20.0)
            super.weight = value
        }

    override var fullHeight: Double
        get() = super.fullHeight
        set(value) {
            assert(fullHeight in 50.0 .. 100.0)
            super.fullHeight = value
        }

    override var fullLength: Double
        get() = super.fullLength
        set(value) {
            assert(fullLength in 50.0 .. 100.0)
            super.fullLength = value
        }
    override var fullWidth: Double
        get() = super.fullWidth
        set(value) {
            assert(fullWidth in 50.0 .. 100.0)
            super.fullWidth = value
        }

    override fun eat(food: Cat): Boolean {
        return if (catEaten.size > 10) {
            println("DOG: ENOUGH")
            false
        } else {
            true
        }
    }

    override fun speak(): String {
        return "DOG: BARK"
    }
}